### PR TITLE
Implement ACP 429: add `Lazy{Cell,Lock}::get[_mut]` and `force_mut`

### PR DESCRIPTION
**Pull request by workingjubilee**
Tracking issue for `lazy_get`: https://github.com/rust-lang/rust/issues/129333

This lands the essential content of https://github.com/rust-lang/rust/pull/129334 without adding DerefMut and without demanding that T-libs-api abrogate their procedure unnecessarily just to land a few functions. The case of the trait implementation can be considered in another PR.

### CODE REVIEW

**Comment by workingjubilee** [comment]
Finally ran down the last missing import.

File: library/core/src/cell/lazy.rs [hunk 1]
```diff
@@ -114,7 +115,72 @@ impl<T, F: FnOnce() -> T> LazyCell<T, F> {
...
+    /// *p = 44;
+    /// assert_eq!(*lazy, 44);
+    /// ```
+    #[inline]
+    #[unstable(feature = "lazy_get", issue = "129333")]
+    pub fn force_mut(this: &mut LazyCell<T, F>) -> &mut T {
+        #[cold]
+        /// # Safety
+        /// May only be called when the state is `Uninit`.
+        unsafe fn really_init<T, F: FnOnce() -> T>(state: &mut State<T, F>) -> &mut T {
```
Comment (Amanieu) [inline library/core/src/cell/lazy.rs:144]: Rename this to `really_init_mut` for clarity. There's already a function called `really_init` just below (yes, I know it's actually under `Self::really_init`, but it's still confusing on a first read).

File: library/std/src/sync/lazy_lock.rs [hunk 2]
```diff
@@ -134,6 +134,60 @@ impl<T, F: FnOnce() -> T> LazyLock<T, F> {
...
+    /// *p = 44;
+    /// assert_eq!(*lazy, 44);
+    /// ```
+    #[inline]
+    #[unstable(feature = "lazy_get", issue = "129333")]
+    pub fn force_mut(this: &mut LazyLock<T, F>) -> &mut T {
+        #[cold]
+        /// # Safety
+        /// May only be called when the state is `Incomplete`.
+        unsafe fn really_init<T, F: FnOnce() -> T>(this: &mut LazyLock<T, F>) -> &mut T {
```
Comment (Amanieu) [inline library/std/src/sync/lazy_lock.rs:159]: Might as well do the same here for consistency.

**Review by Amanieu — COMMENTED** [review]
r=me once comments are addressed.

File: library/core/src/cell/lazy.rs [hunk 3]
```diff
@@ -152,13 +218,55 @@ impl<T, F: FnOnce() -> T> LazyCell<T, F> {
 }
 
 impl<T, F> LazyCell<T, F> {
+    /// Returns a reference to the value if initialized, or `None` if not.
```
Comment (jalil-salame) [inline library/core/src/cell/lazy.rs:221]: **mutable** reference

File: library/core/src/cell/lazy.rs [hunk 4]
```diff
@@ -152,13 +218,55 @@ impl<T, F: FnOnce() -> T> LazyCell<T, F> {
...
+    #[unstable(feature = "lazy_get", issue = "129333")]
+    pub fn get_mut(this: &mut LazyCell<T, F>) -> Option<&mut T> {
+        let state = this.state.get_mut();
+        match state {
+            State::Init(data) => Some(data),
+            _ => None,
+        }
+    }
+
+    /// Returns a mutable reference to the value if initialized, or `None` if not.
```
Comment (jalil-salame) [inline library/core/src/cell/lazy.rs:247]: **non-mutable** reference

**Review by jalil-salame — COMMENTED** [review]
Found some typos

**Pull request merged**